### ISSUE TITLE
Performance of primary Sizzle search

### ISSUE THREAD

**Issue by paulirish** [jquery/sizzle#337]
In recent profiling, I have seen a lot of time being spent in the primary Sizzle function. Obviously the developer isn't caching their objects themselves, but I think we have room to improve on the sizzle side too.

~~Two items~~ One item stood out to me and likely have some easy wins.
### qSA for the simple cases

Edit (aug 2015): removed this one as it was a bad idea. ;)  
### fix the try/catch deopt

This one's much more solid than the first. :)

Because try/catch exists in the body of the function, V8 tosses the entire function into a deoptimized mode, that is orders of magnitude slower. (Thus the warning triangle).  It's my understanding Firefox doesn't have this problem, and some next generation work from V8 will also fix this. So it'll disappear in time,  but in the meantime it's a _very easy_ fix.

Looks something like this: https://gist.github.com/paulirish/cdfce712bba88e264601/revisions
<img src="https://cloud.githubusercontent.com/assets/39191/8021586/f6b2da5c-0c59-11e5-8baa-cd94cc589d73.png" height=300>

**Comment by dmethvin** [comment]
This seems like a good change, would be good to test across all browsers.

**Comment by timmywil** [comment]
The second change seems like a no-brainer. I'm more skeptical about the first. We'd want to do whatever performs the best across all of the browsers, not just v8. Historically, the gEBID and gEBTN shortcuts have been worthwhile.

**Comment by mzgol** [comment]
I believe `qSA('#id')` will be always slower than `gEBID('id')` as the former needs to return all nodes with id `id` even if there are, incorrectly, more than one on the page, whereas the latter can finish the search when it finds the first reference. Sizzle doesn't support documents with repeated ids so we can make such a shortcut.

**Comment by paulirish** [comment]
> I believe qSA('#id') will be always slower than gEBID('id')

yeah seems good. 

was a bad idea.  i'll remove it from the issue above to keep things simple.

**Comment by gibson042** [comment]
This seemed like an easy win, but my testing with Chromium 44 showed negligible effect at best.
```patch
@@ -302,18 +302,10 @@ function Sizzle( selector, context, results, seed ) {
 						context;
 				}
 
-				if ( newSelector ) {
-					try {
-						push.apply( results,
-							newContext.querySelectorAll( newSelector )
-						);
-						return results;
-					} catch ( qsaError ) {
-					} finally {
-						if ( nid === expando ) {
-							context.removeAttribute( "id" );
-						}
-					}
+				if ( newSelector &&
+					tryQSA( newSelector, newContext, results, nid === expando && context ) ) {
+
+					return results;
 				}
 			}
 		}
@@ -323,6 +315,20 @@ function Sizzle( selector, context, results, seed ) {
 	return select( selector.replace( rtrim, "$1" ), context, results, seed );
 }
 
+function tryQSA( selector, context, results, fakeId ) {
+	try {
+		push.apply( results,
+			context.querySelectorAll( selector )
+		);
+		return results;
+	} catch ( qsaError ) {
+	} finally {
+		if ( fakeId ) {
+			fakeId.removeAttribute( "id" );
+		}
+	}
+}
+
 /**
  * Create key-value caches of limited size
  * @returns {function(string, object)} Returns the Object data after storing it on itself with
```

Even though the try/catch marks `Sizzle` as deoptimized, introducing a new call seems to outweigh whatever benefits optimization can bring to such a DOM-heavy function. However, I will keep this open for a little bit in case I missed something.

**Comment by paulirish** [comment]
What are you using to benchmark the perf?

**Comment by gibson042** [comment]
I started with developer tools profiling against a no-try/catch QUnit test suite and a stubbed-out non-qSA fallback... not the most immaculate of environments, but I wanted to subject `Sizzle` to real-ish input. It showed a net _hit_ to total time spent in Sizzle, albeit a largely insignificant one.

Then I just now put together a [jsPerf](http://jsperf.com/sizzle-try-catch/4) to dig a little deeper. It showed no change in qSA-able selectors (as expected), a 43% speedup in the already wicked fast `get*By*`-able simple selectors, and an 8% speedup in the remainder (e.g., custom selectors).

I'm still a bit on the fence, but honestly much less so now than I was when I started writing this reply. Seeing the numbers laid out like that make me think that it might be worth it, even though the savings are small.
